machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,199 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineosconfigs,scope=Cluster
+// +kubebuilder:subresource:status
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1773
+// +openshift:enable:FeatureGate=OnClusterBuild
+// +openshift:file-pattern=cvoRunLevel=0000_80,operatorName=machine-config,operatorOrdering=01
+// +kubebuilder:metadata:labels=openshift.io/operator-managed=
+
+// MachineOSConfig describes the configuration for a build process managed by the MCO
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of the machineosconfig
+	// +kubebuilder:validation:Required
+	Spec MachineOSConfigSpec `json:"spec"`
+
+	// status describes the status of the machineosconfig
+	// +optional
+	Status MachineOSConfigStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSConfigList describes all configurations for image builds on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []MachineOSConfig `json:"items"`
+}
+
+// MachineOSConfigSpec describes user-configurable options as well as information about a build process.
+type MachineOSConfigSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
```

Comment:
Why couldn't a build target more than one pool?

Reply:
this is not in scope for OCL. one config per pool and then multiple build objects per config. Muddying the waters here and opening up the door for 3+ pools to use the same config will make implementation and UX much worse I think

Reply:
OK so its not possible for an OS image to be generic in this context is what your saying.

Reply:
in this scope, yes.

Reply:
and honestly, in the scope of OCL for the foreseeable future. MCO is only in the business of managing images we are going to apply for you not images that just hang out in a registry.